### DISCUSSION TITLE
Kinematic body is lagging behind a Dynamic body

### DISCUSSION THREAD

**Discussion by JoniBife** [jrouwe/JoltPhysics#1295] [Q&A] [answered]
Hi,

I have a dynamic rigibody which is moved by forces (via _AddForce_) then I have a kinematic rigibody which is following the position of the dynamic rigibody by grabbing its position through the bodyinterface and then using _MoveKinematic_ towards target position in delta time (see figure below). Intuitively I can understand that the kinematic body will never reach the rigibody's position as long as it moves, since both are updated simultaneously and therefore the kinematic body is always one physics update behind. Is my intuition correct? Or is there a different problem here? 

![KinematicLag](https://github.com/user-attachments/assets/34c5d420-9f5e-40d5-aece-ba0a0fa10e7f)
*The green box in wireframe represents the dynamic body whereas the yellow one the kinematic*

In an attempt to fix the issue I also checked the ragdoll source code , especifically _DriveToPoseUsingKinematics_ which indeed also uses _MoveKinematic_ to move its joints. This makes me wonder, if I were to play the same animation on top of the Kinematic Rig would it also lag behind? To clarify "lagging behind" I am not referring to the fact that the animation would potentially be running at different frequency than the physics engine but literally lagging behind in both position and rotation. 

My ultimate goal is to have the kinematic body always in the current position of the dynamic body so if this is not the appropriate way of solving such problem how would one do it?

Thanks in advance for the help :)

**Comment by jrouwe** [comment, marked as answer]
> Intuitively I can understand that the kinematic body will never reach the rigibody's position as long as it moves, since both are updated simultaneously and therefore the kinematic body is always one physics update behind. Is my intuition correct?

Yes this is correct.

> This makes me wonder, if I were to play the same animation on top of the Kinematic Rig would it also lag behind?

In general yes, I think there are a couple of options:
1. Accept that kinematic bodies lag behind 1 frame (this is what we do).
2. You can divide your game code into multiple phases: first all the animations would be evaluated and you would call `BodyInterface::MoveKinematic`, then the physics update would run, then the rest of the game logic would run and it would see animation and physics in sync.
3. You can move your kinematic body using `BodyInterface::SetPositionAndRotation`. This works as long as the kinematic body doesn't interact with other dynamic objects. We use a second set of high(er) detail collision volumes around characters that move like this and that are only used for hit detection (bullet ray casts etc.). 
4. You can try to predict the animation pose for the next frame by either extrapolating or by sampling the pose in the future. If this works depends on how predictable the animation is.

> My ultimate goal is to have the kinematic body always in the current position of the dynamic body so if this is not the appropriate way of solving such problem how would one do it?

I think the best solution depends on what the purpose is of that 2nd kinematic body and what interaction it has with the world. It is e.g. possible, through the ContactSettings class, to have the same body act as kinematic against one set of bodies and dynamic for another set of bodies, so you might not need a 2nd body at all.

**Reply by JoniBife** [reply]
Thanks for the quick answer! The second idea seems quite interesting because in my context objects might be moving quite fast therefore that one frame lag becomes exponentially worse and separating the move kinematics this way will likely solve it thanks!
